File: dotnet/src/Rerout/Models/LinkStats.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Rerout.Models;

/// <summary>Click-count breakdown by dimension (country, referrer, etc.).</summary>
public sealed record StatsBreakdown
{
    /// <summary>The bucket — country code, referrer host, device class, etc.</summary>
    [JsonPropertyName("value")]
    public required string Value { get; init; }

    /// <summary>Clicks attributed to this bucket.</summary>
    [JsonPropertyName("clicks")]
    public required long Clicks { get; init; }
}

/// <summary>One day's click + scan totals.</summary>
public sealed record DailyClicksPoint
{
    /// <summary>Unix seconds — start of the day (UTC).</summary>
    [JsonPropertyName("day")]
    public required long Day { get; init; }

    /// <summary>Total clicks recorded on this day.</summary>
    [JsonPropertyName("clicks")]
    public required long Clicks { get; init; }

    /// <summary>Subset of <see cref="Clicks"/> attributed to a QR scan.</summary>
    [JsonPropertyName("qr_scans")]
    public required long QrScans { get; init; }
}

/// <summary>Analytics for a single short link across the requested window.</summary>
public sealed record LinkStats
{
    /// <summary>The short code these stats belong to.</summary>
    [JsonPropertyName("code")]
    public required string Code { get; init; }

    /// <summary>Window size in days the totals span.</summary>
    [JsonPropertyName("days")]
    public required int Days { get; init; }

    /// <summary>Total clicks in the window.</summary>
    [JsonPropertyName("total_clicks")]
    public required long TotalClicks { get; init; }

    /// <summary>Subset of <see cref="TotalClicks"/> attributed to QR scans.</summary>
    [JsonPropertyName("qr_scans")]
    public required long QrScans { get; init; }

    /// <summary>Top countries.</summary>
    [JsonPropertyName("countries")]
    public IReadOnlyList<StatsBreakdown> Countries { get; init; } = [];

    /// <summary>Top referrers.</summary>
    [JsonPropertyName("referrers")]
    public IReadOnlyList<StatsBreakdown> Referrers { get; init; } = [];
}

File: java/src/main/java/co/rerout/sdk/model/LinkStats.java
/*
 * rerout-java — Official Java SDK for the Rerout branded-link API.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk.model;

import com.google.gson.annotations.SerializedName;
import java.util.Collections;
import java.util.List;

/** Analytics for a single short link across the requested window. */
public final class LinkStats {

    private final String code;
    private final int days;

    @SerializedName("total_clicks")
    private final long totalClicks;

    @SerializedName("qr_scans")
    private final long qrScans;

    private final List<StatsBreakdown> countries;
    private final List<StatsBreakdown> referrers;

    /**
     * Creates a {@code LinkStats}.
     *
     * @param code        the short code these stats belong to
     * @param days        the window size in days the totals span
     * @param totalClicks total clicks in the window
     * @param qrScans     the subset of {@code totalClicks} attributed to a QR scan
     * @param countries   top countries by click count
     * @param referrers   top referrers by click count
     */
    public LinkStats(
            String code,
            int days,
            long totalClicks,
            long qrScans,
            List<StatsBreakdown> countries,
            List<StatsBreakdown> referrers) {
        this.code = code;
        this.days = days;
        this.totalClicks = totalClicks;
        this.qrScans = qrScans;
        this.countries = countries;
        this.referrers = referrers;
    }

    /** {@return the short code these stats belong to} */
    public String getCode() {
        return code;
    }

    /** {@return the window size in days the totals span} */
    public int getDays() {
        return days;
    }

    /** {@return total clicks in the window} */
    public long getTotalClicks() {
        return totalClicks;
    }

    /** {@return the subset of {@link #getTotalClicks()} attributed to a QR scan} */
    public long getQrScans() {
        return qrScans;
    }

    /** {@return top countries by click count; never {@code null}} */
    public List<StatsBreakdown> getCountries() {
        return countries == null ? Collections.emptyList() : countries;
    }

    /** {@return top referrers by click count; never {@code null}} */
    public List<StatsBreakdown> getReferrers() {
        return referrers == null ? Collections.emptyList() : referrers;
    }
}
